### ISSUE TITLE
OpenSSL 1.1.0 API Changes  (was "Failing on Pthread compilation")

### ISSUE THREAD

**Issue by GrahamMThomas** [emptymonkey/revsh#10]
![xvsvb9fr](https://cloud.githubusercontent.com/assets/14031456/25930035/9901b826-35b8-11e7-94c0-9a129c5c9511.jpg)

Having compilation errors that appear to have a problem with the pthread library. Issue occurs using the top two build options inside the Makefile. 

Issue has been seen on multiple Kali boxes, all of which are completely updated.

```
root@kali:~/Desktop/revsh# uname -a
Linux kali 4.9.0-kali4-amd64 #1 SMP Debian 4.9.25-1kali1 (2017-05-04) x86_64 GNU/Linux
```

Can get passed this error by adding -pthread to the LIBS in the `Makefile`. After which, I am confronted with this error:

![virtualbox_2017-05-10_19-27-38](https://cloud.githubusercontent.com/assets/14031456/25930113/26b28768-35b9-11e7-8359-3f43cb39ecf4.png)

**Comment by emptymonkey** [comment]
Confirmed. Several issues with building in Kali. One of these appears to be a change in the openssl command line parsing. Other issues seem to be related to kali libraries. I'll work on this and get it cleaned up.

**Comment by emptymonkey** [comment]
Changed issue name. Everything here is because kali uses the newer OpenSSL 1.1. There are quite a few changes to the api. This may take some time, but is being worked on.

**Comment by mzpqnxow** [comment]
@emptymonkey are you sticking with openssl? Do you have any interest in switching to mbedtls and dropping openssl altogether? I would be happy to help port to mbedtls if I have some spare time as I had intended to put something like this together already, but it seems you've done most of the work for me..

https://github.com/ARMmbed/mbedtls

**Comment by emptymonkey** [comment]
@mzpqnxow , I'm not set on any library. My only real design goals with revsh are:
* The config and building isn't too painful for the end user.
* everything "just works" once built. (This implies a full static build to rule out compat issues on the target. This is where OpenSSL 1.1 is failing pretty badly these days.)

I've been considering both mbedtls and wolfssl. (I'm not partial to one or the other. I've only ever coded with OpenSSL and am happy for the change.)  Unfortunately, life has gotten in the way of code right now, so I've been saying I'd work on this for several months but honestly haven't started. I'm happy for any help people want to contribute. 

As for the mechanics, revsh is pretty modular right now. Other than a small number of "#ifdef OPENSSL" entries, everything of real interest is in the io_ssl.c file. The path forward I see in my head is to rename that file as io_openssl.c and then add an io_mbedtls.c, etc. All interchangeable depending on how the end user wants to build it.

Finally, because I have a bit of a fetish for tight binaries, I'm also interested in building both the tls lib and revsh itself with an embedded c lib like musl libc once we get the new tls lib code added.  Hit me up on email or twitter dm if you want to discuss how the internals of revsh are put together. (Start with Documentation/CODEMAP. Oh, and I have a devel branch where I smash it all together first, since lots of people seem to be using this these days I try not to jump straight to master.) Take a look under the hood. If it still sounds like something you want to work on, I'd be happy for the help. Cheers!

**Comment by mzpqnxow** [comment]
Thanks- yeah I will fork and let you know if I start putting any serious effort into getting away from OpenSSl- and yes, musl is a good option, I've already built a few statically linked revsh bins for a few platforms using musl-cross-make which makes it very painless (vi config.mak, make -j && make install)

I have configs for a few different platforms @ https://github.com/mzpqnxow/musl-cross-make-configs if you have any interest in messing with musl-cross-make, just replace config.mak with one of those configs, make clean && make -j && make install. Check it out if you haven't already and want to see how streamlined musl-cross-make is. Building the musl toolchain, openssl static libs and revsh statically linked for MIPS-I only took me about 15 mins, which isn't all that bad.

**Comment by emptymonkey** [comment]
I'm going to go ahead and close this out as I seem to have it working now with the OpenSSL 1.1 family. There have been enough fixes to OpenSSL 1.1 that it now builds and works with a static binary, but only if you ignore the libssl.a and libcrypto.a files that ship with the distros and build from more recent source. I've updated the Install section of the README.md file as well as the Makefile to reflect this (in devel branch for now). 

I haven't done a lot of testing yet. Once I make some more changes / do some more cleaning, I'll do a round of testing and push everything to master. Thanks everyone for the help!
